lib/Fhp/Action/SendSEPATransfer.php

```diff
@@ -35,21 +47,52 @@ class SendSEPATransfer extends BaseAction
      */
     public static function create(SEPAAccount $account, string $painMessage): SendSEPATransfer
     {
-        if (preg_match('/xmlns="(.*?)"/', $painMessage, $match) === false) {
+        if (preg_match('/xmlns="(.*?)"/', $painMessage, $xmlns) === false) {
             throw new \InvalidArgumentException('xmlns not found in the PAIN message');
         }
+
+        // Check whether the PAIN message contains the correct message root
+        if (strstr($painMessage, '<CstmrCdtTrfInitn>') === false) {
+            throw new \InvalidArgumentException('Pain message contains a wrong message root');
+        }
+
+        // Check whether the PAIN message contains multiple or only one transfer, should match <NbOfTxs>xx</NbOfTxs> in the XML
+        $nbOfTxs = substr_count($painMessage, '<CdtTrfTxInf>');
+        if ($nbOfTxs === 0) {
+            throw new \InvalidArgumentException('PAIN message contains no credit transfer transactions');
+        }
+
+        $ctrlSum = null;
+        if (preg_match('@<GrpHdr>.*<CtrlSum>(?<ctrlsum>[.0-9]+)</CtrlSum>.*</GrpHdr>@s', $painMessage, $matches) === 1) {
+            $ctrlSum = $matches['ctrlsum'];
+        }
+
+        // Check whether a <PmtInf> block sets <BtchBookg> to false: This means that Single Booking is to be requested
+        $singleBooking = preg_match('@<PmtInf>.*<BtchBookg>false</BtchBookg>.*</PmtInf>@s', $painMessage) === 1;
+
+        // Set RequestedExecutionDate to 1999-01-01, as required by the FinTS standard for non-scheduled transfers
+        $painMessage = preg_replace('@(<ReqdExctnDt>)([\d-]*)(</ReqdExctnDt>)@', '${1}1999-01-01${3}', $painMessage) ?? $painMessage;
```

Comment:
Should the date updating be wrapped in `if (!$result->singleTransfer)`?

Reply:
Also maybe we shouldn't have this date updating at all and just reject all other dates? Otherwise the library user believes they're getting a particular date, but the library silently drops it, which would be unexpected.

Reply:
No, the date should be 1999-01-01 also for singleTransfers. Up to now, those files would be rejected by the bank. We use files from external sources with conflicting dates. and the specification [explicitly says (page 496)](https://www.hbci-zka.de/dokumente/spezifikation_deutsch/fintsv3/FinTS_3.0_Messages_Geschaeftsvorfaelle_2015-08-07_final_version.pdf) 

> Wird die SEPA-pain-message aus einer fremden Quelle importiert, so ist darauf zu achten, dass ein eventuell abweichendes Datum im Feld RequestedExecutionDate <ReqdExctnDt> auf den Wert „1999-01-01“ geändert wird

(Translation by me: If the SEPA pain message is imported from an external source, one should take care to change a potentially differing date in the field RequestedExecutionDate <ReqdExctnDt> to "1999-01-01".)

That's exactly what I do here.

Actually, this concern should be solved by issuing the as-yet-unimplemented HKCSE/HKCME segments for _scheduled_ transfers. Which I want to do in the near future, since we will need those, too, on our side. For now, I will add a TODO comment to that effect.

Reply:
I see, thanks for the explanation.